source/mesh_deformation/free_surface.cc

```diff
@@ -307,15 +307,27 @@ namespace aspect
 
                 cell_vector = 0;
                 cell_matrix = 0;
+                Tensor<1,dim> direction;
+                if (advection_direction == SurfaceAdvection::geometric_normal)
+                  // Only for 2D
+                  {
+                    Tensor<1,dim> tangent = (fs_fe_face_values.quadrature_point(n_face_q_points-1) - fs_fe_face_values.quadrature_point(0));
+                    direction[0] = -tangent[1];
+                    direction[1] = tangent[0];
+                    // Check that normal points upwards
+                    if (direction[1] < 0)
+                      direction *= -1.;
+                  }
                 for (unsigned int point=0; point<n_face_q_points; ++point)
                   {
                     // Select the direction onto which to project the velocity solution
-                    Tensor<1,dim> direction;
                     if ( advection_direction == SurfaceAdvection::normal ) // project onto normal vector
-                      direction = fs_fe_face_values.normal_vector(point);
+                      {
+                        direction = fs_fe_face_values.normal_vector(point);
+                      }
                     else if ( advection_direction == SurfaceAdvection::vertical ) // project onto local gravity
                       direction = this->get_gravity_model().gravity_vector(fs_fe_face_values.quadrature_point(point));
-                    else
+                    else if (advection_direction != SurfaceAdvection::geometric_normal)
```

Comment:
now that you have 3 cases, it would probably be worthwhile making this into a `switch...case` thing